### PR TITLE
SNOW-1651234: Fix create_dataframe throwing an exception for structured dtypes

### PR DESCRIPTION
**Pull request by sfc-gh-jrose**
1. Which Jira issue is this PR addressing? Make sure that there is an accompanying issue to your PR.

   

   Fixes SNOW-1651234

2. Fill out the following pre-review checklist:

   - [ ] I am adding a new automated test(s) to verify correctness of my new code
      - [ ] If this test skips Local Testing mode, I'm requesting review from @snowflakedb/local-testing
   - [ ] I am adding new logging messages
   - [ ] I am adding a new telemetry message
   - [ ] I am adding new credentials
   - [ ] I am adding a new dependency
   - [ ] If this is a new feature/behavior, I'm adding the Local Testing parity changes.

3. Please describe how your code solves the related issue.

   This change modifies the logic for `session.create_dataframe` to add support for StructType object as well as removes it's dependency on `to_object` and `to_array`. Those calls create semi-structured columns which are incompatible with iceberg. By switching to a cast full type information is retained instead.

### CODE REVIEW

File: tests/integ/scala/test_datatype_suite.py [hunk 1]
```diff
@@ -469,29 +458,54 @@ def test_structured_dtypes_iceberg(
...
+    structured_type_session, local_testing_mode, structured_type_support
+):
+    if not (
+        structured_type_support
+        and iceberg_supported(structured_type_session, local_testing_mode)
+    ):
+        pytest.skip("Test requires iceberg support and structured type support.")
+
+    _, __, expected_schema = STRUCTURED_TYPES_EXAMPLES[True]
+    table_name = f"snowpark_structured_dtypes_{uuid.uuid4().hex[:5]}"
```
Comment (sfc-gh-aalam) [inline tests/integ/scala/test_datatype_suite.py:489]: just curious, any reason why we are using this naming format instead os standard function `Utils.random_table_name()`?

**Review by sfc-gh-aalam — APPROVED** [review]

File: tests/integ/scala/test_datatype_suite.py (hunk 1 above)
Comment (sfc-gh-jrose) [reply tests/integ/scala/test_datatype_suite.py:489]: No particular reason. Other tests in this file did it this way so I was mostly just copying style.

**Comment by sfc-gh-jrose** [comment]
Defering this change until structured type support is rolled out completely.

File: CHANGELOG.md [hunk 2]
```diff
@@ -49,6 +49,7 @@
 - Fixed a bug where calling `DataFrame.to_snowpark_pandas_dataframe` without explicitly initializing the Snowpark pandas plugin caused an error.
 - Fixed a bug where using the `explode` function in dynamic table creation caused a SQL compilation error due to improper boolean type casting on the `outer` parameter.
 - Fixed a bug where using `to_pandas_batches` with async jobs caused an error due to improper handling of waiting for asynchronous query completion.
+- Fixed a bug in `session.create_dataframe` that caused a sql error when creating an iceberg table with structured datatypes.
```
Comment (sfc-gh-yixie) [inline CHANGELOG.md:52]: Better remove this one because there are other problems with create_dataframe.

**Comment by sfc-gh-yixie** [comment]
We can continue to merge this PR since it doesn't hurt anything.

**Review by sfc-gh-yixie — APPROVED** [review]

**Pull request merged**